spec/models/article_spec.rb

```diff
@@ -32,6 +32,7 @@ def build_and_validate_article(*args)
     it { is_expected.to have_many(:tags) }
     it { is_expected.to have_many(:user_subscriptions).dependent(:nullify) }
 
+    it { is_expected.to validate_length_of(:body_markdown).is_at_least(0).is_at_most(800.kilobytes) }
```

Comment:
how does this work with utf8 encoding? If I create an article 750,000 Chinese characters using multibyte entities, would length be in characters (unicode entities) or bytes? Is 3MB a safe length?

```
irb(main):005:0> ('𑨕' * 800 * 1024).length
=> 3276800
```
Just choosing Zanabazar square letter ttha as an example (0x11a15) here I get perhaps more than you anticipated.

If the issue with the database is _also_ expressed in characters and not bytes, feel free to dismiss this!

Reply:
Good point! The DB says `max 1048575 bytes`. From https://www.postgresql.org/docs/11/textsearch-limitations.html

> The length of a tsvector (lexemes + positions) must be less than 1 megabyte

Rails uses `string.length` to check https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/activemodel/lib/active_model/validations/length.rb#L41

Reply:
is there a "validate bytesize of" function?

Reply:
@rhymes That'd be characters (Unicode codepoints). If we want bytes, we need to use `String#bytesize`.

Reply:
Also, why do we have specs on validations and associations like this? Is there any way to make this pass other than the exact line of code for the validation?

Reply:
> That'd be characters (Unicode codepoints). If we want bytes, we need to use String#bytesize.

Ok we need a custom validator then. I think this is why Rails doesn't have a bytesize validator: https://github.com/rails/rails/pull/22870#issuecomment-169439090

We can use something like this https://github.com/rails/rails/issues/19570#issuecomment-348366536

> Also, why do we have specs on validations and associations like this? Is there any way to make this pass other than the exact line of code for the validation?

I'm not sure I understand this part 😊

Reply:
For some reason I really thought Rails was using `.bytesize` internally 😭

Reply:
> > Also, why do we have specs on validations and associations like this? Is there any way to make this pass other than the exact line of code for the validation?
> 
> I'm not sure I understand this part 😊

As far as I know, the only way to make `expect(Article).to validate_length_of(:body_markdown)...` pass is to have `validates :body_markdown, length: { ... }` in `Article`. It specifies the implementation of the validation rather than the interface.

We could test the interface by saying something like this:

```ruby
article = build(:article)

article.body_markdown = '.' * 800.kilobytes
expect(article).to be_valid

article.body_markdown << '.'
expect(article).to be_invalid
expect(article.errors[:body_markdown]).to include "is too long"

article.body_markdown = '❤️' * (800.kilobytes / '❤️'.bytesize) # 819_198 out of 819_200 allowed bytes
expect(article).to be_valid

article.body_markdown << '❤️' # 819_204 bytes, exceeds 819_200 allowed
expect(article).not_to be_valid
expect(article.errors[:body_markdown]).to include "is too long"
```

This way, it doesn't matter how we change the implementation of the validation as long as it conforms to the invariant.

Reply:
Ah, got it! 

I think we're covered. Right now we have a length validator:

https://github.com/forem/forem/blob/9fe7be190bd364f48deb8895f845378a2c9bfa27/app/models/article.rb#L57

which is now tested by 

```ruby
it { is_expected.to validate_length_of(:body_markdown).is_at_least(0) }
```

and after the latest change we'll also have a `bytesize` validator and its test